[ASP-2755] Update documentation (#231)
omnivector-solutions/license-manager: docs-source/src/architecture.rst

License Manager Agent
---------------------
The ``License Manager Agent`` is responsible for keeping the local cluster license totals
in sync with the the 3rd party license server totals. It's also responsible for making booking requests
to the ``License Manager API`` when Slurm is configured to use the ``PrologSlurmctld`` script provided by ``License Manager Agent``.

Reconciliation
**************

> For each license tracked by License Manager, the ``License Manager Agent`` will periodically poll the license server to get

Reviewer: Maybe `servers` or `server(s)`?

```suggestion
For each license tracked by License Manager, the ``License Manager Agent`` will periodically poll the license servers to get
```

Author: Fixed!

the usage information and store it in the ``License Manager API``. The ``stat-interval`` is the period of time
between each reconciliation and can be configured in the ``License Manager Agent`` configuration file.

The information in the ``License Manager API`` is used by the reconciliation process to update the license counters in Slurm.
This is done by creating a reservation to represent the licenses used in the license server.

This reservation is not meant to be consumed by users nor jobs; it's only a representation of the licenses in use.
The reservation is created by the user configured in the ``License Manager Agent`` configuration file. The user must
have a user account in the Slurm cluster and have ``operator`` privilege level to manage reservations.

Bookings
********
The ``License Manager Agent`` is also responsible for making booking requests to the ``License Manager API``
when Slurm is configured to use the ``PrologSlurmctld`` script provided by ``License Manager Agent``.

> Each job submitted to Slurm will make a request to the ``License Manager API`` to book the needed licenses prior

Reviewer: I think it would be good to clarify that ``License Manager Agent`` makes the request to the API.

Author: ```suggestion
Each job submitted to Slurm will trigger the ``PrologSlurmctld`` script that makes a request to the ``License Manager API``
```
Is this better?

Reviewer: Yes

After the ``grace time`` expires, the booking is deleted. This means that the license was checked out from the
license server and doesn't need a booking anymore.

License Manager API
-------------------

> The ``License Manager API`` provides a RESTful API where licenses and bookins are tracked.

Reviewer: Maybe cite that the CLI interacts with it as well?

Author: ```suggestion
The ``License Manager API`` provides a RESTful API where licenses and bookins are tracked.
The ``License Manager Agent`` uses this API to store the license usage information and to process the booking requests.
The ``License Manager CLI`` interacts with this API to add new configurations and to check the usage information for each tracked license.
```

Better?

Reviewer: Yes

The ``License Manager Agent`` uses this API to store the license usage information and to process the booking requests.

The API is also responsible for verifying if the booking requests can be satisfied by accounting for bookings already
made and the license usage in the license server.

The API contains 8 entities that have relationship among them.
This means that some of the resources need to be created before others can be created as well.

> .. image:: images/diagram.png

Reviewer: Do you see a way to refer to where this diagram can be edited? Maybe adding the link to its source as a comment in this markdown file, or adding the source file (like SVG for others) to the repo as well. In this way, it will be easier to update it if needed.

As a side note, many documentation frameworks support [Mermaid](https://mermaid.js.org). It is an excellent way to have the diagram as a code and track changes to it on version control. On the other hand, it can be difficult to fine-tune its design, so sometimes it is not the ideal tool. For instance, see the diagram for the new data model on Jobbergate: https://gist.github.com/fschuch/51006fa5dc18ab24dee8533bf3fefe48

Author: This is great suggestion, I looked into adding Mermaid into Sphinx and used it to do a new version of the diagram. Thank you!
